Perform loop unroll optimization

nkrkv opened the pull request:
Closes #903

### The outcome

For patch depicted in #903 (Flash / RAM / Transactions per second):

* Before: 4244 (13%) / 1030 (50%) / 34
* After: 4310 (13%) / 147 (7%) / 37
* Comparison: 1.5% decline / 600% improvement / 9% improvement

Note, the TPS measurement methodic is very poor. Most of the time the MCU spends for sending text through serial, not to perform the useful work. So the real improvement should be way greater than 9%.

The patch under the test is somewhat artificial. Trying the new updated transpiller on a real-world SD data logger example has led to following:

* Before: 23588 (73%) / 2828 (138%)
* After: 26204 (81%) / 1903 (92%)
* Comparison: 10% decline / 49% improvement (and, yes, it finally fits Arduino Uno)

### Review

PR is large. I recommend reviewing the commit after commit, they are squashed by scope.

The change touches every program generated by XOD greatly. I have tested the new implementation on a dozen of pocket projects and they seem to work well. However, dear reviewers, if you have any suspicions, edge case ideas, etc, please check and let me know.

brusherru commented:
🐛 
Tried to upload blink project with `watch` node  in the debug mode produces error on transpilation:
```
Error occured during uploading: ChildProcessError: Command failed: "/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/arduino-builder" -hardware="/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/hardware" -hardware="/Users/annies/Library/Application Support/xod-client-electron/packages" -libraries="/Users/annies/Sites/xod/packages/xod-client-electron/arduino-libraries" -tools="/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/tools" -tools="/Users/annies/Library/Application Support/xod-client-electron/packages" -fqbn="arduino:avr:uno" -build-path="/Users/annies/Library/Application Support/xod-client-electron/upload-temp/build" "/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp"
/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp: In function 'void xod::xod__core__watch::evaluate(xod::xod__core__watch::Context)':
/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:1067:25: error: call of overloaded 'print(xod::xod__core__watch::ContextObject*&)' is ambiguous
DEBUG_SERIAL.print(ctx);
^
/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:1067:25: note: candidates are:
In file included from /Users/annies/Library/Application Support/xod-client-electron/packages/arduino/hardware/avr/cores/arduino/Stream.h:26:0,
from /Users/annies/Library/Application Support/xod-client-electron/packages/arduino/hardware/avr/cores/arduino/HardwareSerial.h:29,
from /Users/annies/Library/Application Support/xod-client-electron/packages/arduino/hardware/avr/cores/arduino/Arduino.h:232,
from /Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:18:
/Users/annies/Library/Application Support/xod-client-electron/packages/arduino/hardware/avr/cores/arduino/Print.h:68:12: note: size_t Print::print(char) <near match>
size_t print(char);
```

nkrkv commented:
@brusherru Ooops. Fixed.

TL;DR

The `nodeId` concept has vanished and the `Context` is no longer a Node ID. `watch` is the only node (at the moment) who wants to know own ID. So I added another `#pragma` feature to enable access to the node id if someone really needs it. The `watch` node is the only user for now.

See the last fixup commit. Will squash before the merge.
